aten/src/ATen/native/LinearAlgebra.cpp

```diff
@@ -1908,7 +1911,84 @@ Tensor& vdot_out(const Tensor& self, const Tensor& other, Tensor& result) {
   return result.fill_(self.vdot(other));
 }
 
-static bool should_fold(const Tensor& tensor1, const Tensor& tensor2) {
+#define AUTOTUNE_MAX_DIM 4
+
+struct MatmulAutotuneCacheKey {
+  int tensor1_dim[AUTOTUNE_MAX_DIM];
+  int tensor1_stride[AUTOTUNE_MAX_DIM];
+  int tensor2_dim[AUTOTUNE_MAX_DIM];
+  int tensor2_stride[AUTOTUNE_MAX_DIM];
+  bool should_fold;
+  ScalarType dtype;
+  bool is_cuda;
+  int device;
+};
+
+struct MatmulAutotuneCacheKeyWrapper : ParamsWrapper<MatmulAutotuneCacheKey> {
+  MatmulAutotuneCacheKeyWrapper() = default;
+
+  MatmulAutotuneCacheKeyWrapper(
+    const Tensor& tensor1,
+    const Tensor& tensor2,
+    bool should_fold) {
+    TORCH_INTERNAL_ASSERT(tensor1.sizes().size() <= AUTOTUNE_MAX_DIM);
```

Comment:
TODO: fail gracefully here (bail out of autotuning rather than crashing)

Reply:
Using a Dimvector should do here

Reply:
(DimVector unfortunately doesn't seem to work at it isn't POD, a requirement of the hashing util)